File: playgrounds/pymongo/app/models/book.py
"""Example "book" collection helpers for the PyMongo demo.

PyMongo is a **raw driver**, not an ODM: there is no schema class. Instead this
module centralizes everything the app and tests need to treat a plain BSON
document as a "Book":

  - the collection name,
  - the index definitions (no ``collation`` — DocumentDB does not implement it),
  - a builder that applies defaults + timestamps to a create payload, and
  - a JSON serializer that makes ``_id``/``datetime`` values response-friendly.

It mirrors the Mongoose/Beanie ``Book`` models in the sibling playgrounds so the
three demos are directly comparable.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Mapping

from pymongo import ASCENDING, IndexModel

COLLECTION_NAME = "books"

# Compound index: exercises DocumentDB index creation via PyMongo.
# No `collation` option here; DocumentDB does not implement it.
INDEXES = [
    IndexModel([("author", ASCENDING), ("title", ASCENDING)], name="author_title"),
]


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def build_document(payload: Mapping[str, Any]) -> dict:
    """Apply defaults + timestamps to a create payload, returning a new dict."""
    now = _utcnow()
    return {
        "title": payload["title"],
        "author": payload["author"],
        "genres": list(payload.get("genres", [])),
        "pages": payload.get("pages"),
        "published": payload.get("published"),
        "in_stock": payload.get("in_stock", True),
        "rating": payload.get("rating"),
        "created_at": now,
        "updated_at": now,
    }


def serialize(doc: Mapping[str, Any] | None) -> dict | None:
    """Convert a BSON document into a JSON-serializable dict."""
    if doc is None:
        return None
    out = dict(doc)
    if "_id" in out:
        out["_id"] = str(out["_id"])
    for key in ("published", "created_at", "updated_at"):
        value = out.get(key)
        if isinstance(value, datetime):
            out[key] = value.isoformat()
    return out

File: playgrounds/spring-data-mongodb/app/src/main/java/com/example/playground/model/Book.java
package com.example.playground.model;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.mapping.Document;

/**
 * Example "book" document, mirroring the Mongoose/Beanie/PyMongo playgrounds so
 * the demos are directly comparable.
 *
 * <p>The compound index exercises DocumentDB index creation (via Spring Data
 * {@code auto-index-creation}). No {@code collation} is declared; DocumentDB
 * does not implement collation indexes.
 */
@Document(collection = "books")
@CompoundIndex(name = "author_title", def = "{'author': 1, 'title': 1}")
public class Book {

    @Id
    private String id;

    private String title;
    private String author;
    private List<String> genres = new ArrayList<>();
    private Integer pages;
    private Instant published;
    private Boolean inStock = Boolean.TRUE;
    private Integer rating;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public List<String> getGenres() {
        return genres;
    }

    public void setGenres(List<String> genres) {
        this.genres = genres;
    }

    public Integer getPages() {
        return pages;
    }

    public void setPages(Integer pages) {
        this.pages = pages;
    }

    public Instant getPublished() {
        return published;
    }

    public void setPublished(Instant published) {
        this.published = published;
    }

    public Boolean getInStock() {
        return inStock;
    }

    public void setInStock(Boolean inStock) {
        this.inStock = inStock;
    }

    public Integer getRating() {
        return rating;
    }

    public void setRating(Integer rating) {
        this.rating = rating;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Instant updatedAt) {
        this.updatedAt = updatedAt;
    }
}
